File: web/types/agent-llm-config.ts
/**
 * Types for Agent LLM Configuration
 */

export type RoutingMode = 'fixed' | 'round_robin' | 'cost_opt' | 'intent' | 'latency_opt'

export const ROUTING_MODE_LABELS: Record<RoutingMode, string> = {
  fixed: 'Fixed',
  round_robin: 'Round Robin',
  cost_opt: 'Cost Optimized',
  intent: 'Intent-Based',
  latency_opt: 'Latency Optimized',
}

export const ROUTING_MODE_DESCRIPTIONS: Record<RoutingMode, string> = {
  fixed: 'Always use the default model. No routing overhead.',
  round_robin: 'Distribute requests across models by weight. Good for load balancing and A/B testing.',
  cost_opt: 'Use the cheapest model that can handle the query complexity. Best for cost reduction.',
  intent: 'Match the model to the query type (code, math, creative, etc.). Best for quality per task.',
  latency_opt: 'Always prefer the fastest model. Best for real-time chat.',
}

export const INTENT_OPTIONS = [
  { value: 'code', label: 'Code' },
  { value: 'math', label: 'Math' },
  { value: 'research', label: 'Research' },
  { value: 'creative', label: 'Creative Writing' },
  { value: 'data_analysis', label: 'Data Analysis' },
  { value: 'simple_qa', label: 'Simple Q&A' },
  { value: 'reasoning', label: 'Complex Reasoning' },
  { value: 'general', label: 'General' },
]

export interface RoutingRules {
  intents?: string[]
  min_complexity?: number
  max_complexity?: number
  cost_per_1k_input?: number
  cost_per_1k_output?: number
  priority?: number
  is_fallback?: boolean
}

export interface AgentLLMConfig {
  id: string;
  agent_id: string;
  tenant_id: string;
  name: string;
  provider: string;
  model_name: string;
  api_base?: string;
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  additional_params?: Record<string, any>;
  is_default: boolean;
  display_order: number;
  enabled: boolean;
  routing_rules?: RoutingRules;
  routing_weight?: number;
  created_at: string;
  updated_at: string;
}

export interface AgentLLMConfigCreate {
  name: string;
  provider: string;
  model_name: string;
  api_key: string;
  api_base?: string;
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  additional_params?: Record<string, any>;
  is_default?: boolean;
  display_order?: number;
  enabled?: boolean;
  routing_rules?: RoutingRules;
  routing_weight?: number;
}

export interface AgentLLMConfigUpdate {
  name?: string;
  provider?: string;
  model_name?: string;
  api_key?: string;
  api_base?: string;
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  additional_params?: Record<string, any>;
  is_default?: boolean;
  display_order?: number;
  enabled?: boolean;
  routing_rules?: RoutingRules;
  routing_weight?: number;
}

export interface AgentLLMConfigReorder {
  config_orders: Array<{
    config_id: string;
    display_order: number;
  }>;
}

// Note: Provider and model lists are now fetched dynamically from the API
// See web/lib/api/llm-providers.ts for API client functions

/**
 * Helper function to get a readable model label
 * @param provider - The LLM provider name
 * @param modelName - The model identifier
 * @returns A formatted model label
 */
export function getModelLabel(provider: string, modelName: string): string {
  // Return the model name as-is, or format it if needed
  return modelName || 'Unknown Model'
}

File: api/src/schemas/agent_llm_config.py
"""
Pydantic schemas for agent LLM configurations.
"""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class AgentLLMConfigBase(BaseModel):
    """Base schema for agent LLM configuration."""

    name: str = Field(..., description="Display name for the LLM config")
    provider: str = Field(..., description="LLM provider (google, openai, anthropic, litellm)")
    model_name: str = Field(..., description="Model name")
    api_base: str | None = Field(None, description="API base URL for LiteLLM")
    temperature: float | None = Field(0.7, description="Temperature setting")
    max_tokens: int | None = Field(None, description="Maximum tokens")
    top_p: float | None = Field(None, description="Top P setting")
    additional_params: dict | None = Field(default_factory=dict, description="Additional parameters")
    is_default: bool = Field(False, description="Whether this is the default config")
    display_order: int = Field(0, description="Display order")
    enabled: bool = Field(True, description="Whether this config is enabled")
    routing_rules: dict[str, Any] | None = Field(
        None,
        description=(
            "Routing rules for this config: intents (list of intent tags), "
            "min_complexity (0.0-1.0), max_complexity (0.0-1.0), "
            "cost_per_1k_input (USD), cost_per_1k_output (USD), "
            "priority (int, lower=preferred), is_fallback (bool)"
        ),
    )
    routing_weight: float | None = Field(
        None, description="Weight for round_robin routing (0.0-1.0, default 1.0)"
    )


class AgentLLMConfigCreate(AgentLLMConfigBase):
    """Schema for creating an agent LLM configuration."""

    api_key: str = Field(..., description="API key for the LLM provider")


class AgentLLMConfigUpdate(BaseModel):
    """Schema for updating an agent LLM configuration."""

    name: str | None = Field(None, description="Display name for the LLM config")
    provider: str | None = Field(None, description="LLM provider")
    model_name: str | None = Field(None, description="Model name")
    api_key: str | None = Field(None, description="API key for the LLM provider")
    api_base: str | None = Field(None, description="API base URL for LiteLLM")
    temperature: float | None = Field(None, description="Temperature setting")
    max_tokens: int | None = Field(None, description="Maximum tokens")
    top_p: float | None = Field(None, description="Top P setting")
    additional_params: dict | None = Field(None, description="Additional parameters")
    is_default: bool | None = Field(None, description="Whether this is the default config")
    display_order: int | None = Field(None, description="Display order")
    enabled: bool | None = Field(None, description="Whether this config is enabled")
    routing_rules: dict[str, Any] | None = Field(None, description="Routing rules for this config")
    routing_weight: float | None = Field(None, description="Weight for round_robin routing")


class AgentLLMConfigResponse(AgentLLMConfigBase):
    """Schema for agent LLM configuration response."""

    id: UUID
    agent_id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AgentLLMConfigReorder(BaseModel):
    """Schema for reordering LLM configurations."""

    config_orders: list[dict] = Field(
        ...,
        description="List of config IDs and their new display orders",
        json_schema_extra={"example": [{"id": "uuid-here", "display_order": 0}]},
    )
